internal/provider/common/role.go

```diff
@@ -0,0 +1,170 @@
+	} else if scopeType == "workspace" {
+		roles = workspaceRoles
+	} else if scopeType == "deployment" {
+		nonDeploymentRoles := append(organizationRoles, workspaceRoles...)
+		return !lo.Contains(nonDeploymentRoles, role)
+	}
+
+	return lo.Contains(roles, role)
+}
+
+type ValidateWorkspaceDeploymentRolesInput struct {
+	PlatformClient  *platform.ClientWithResponses
+	OrganizationId  string
+	DeploymentRoles []iam.DeploymentRole
+	WorkspaceRoles  []iam.WorkspaceRole
+}
+
+// ValidateWorkspaceDeploymentRoles checks if deployment roles have corresponding workspace roles
+func ValidateWorkspaceDeploymentRoles(ctx context.Context, diags diag.Diagnostics, input ValidateWorkspaceDeploymentRolesInput) diag.Diagnostics {
+	// return nil if there are no deployment roles
+	if len(input.DeploymentRoles) == 0 {
+		return nil
+	}
+
+	// get list of deployment ids
+	deploymentIds := lo.Map(input.DeploymentRoles, func(role iam.DeploymentRole, _ int) string {
+		return role.DeploymentId
+	})
+
+	// get list of deployments
+	listDeployments, err := input.PlatformClient.ListDeploymentsWithResponse(ctx, input.OrganizationId, &platform.ListDeploymentsParams{
+		DeploymentIds: &deploymentIds,
+	})
+	if err != nil {
+		tflog.Error(ctx, "failed to mutate Team roles", map[string]interface{}{"error": err})
+		diags.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to mutate Team roles and list deployments, got error: %s", err),
+		)
+		return diags
+	}
+	_, diagnostic := clients.NormalizeAPIError(ctx, listDeployments.HTTPResponse, listDeployments.Body)
+	if diagnostic != nil {
+		diags.Append(diagnostic)
+		return diags
+	}
+
+	// get list of workspace ids from deployments
+	deploymentWorkspaceIds := lo.Map(listDeployments.JSON200.Deployments, func(deployment platform.Deployment, _ int) string {
+		return deployment.WorkspaceId
+	})
+
+	// get list of workspaceIds
+	workspaceIds := lo.Map(input.WorkspaceRoles, func(role iam.WorkspaceRole, _ int) string {
+		return role.WorkspaceId
+	})
+
+	// check if deploymentWorkspaceIds are in workspaceIds
+	workspaceIds = lo.Intersect(lo.Uniq(workspaceIds), lo.Uniq(deploymentWorkspaceIds))
+	if len(workspaceIds) != len(deploymentWorkspaceIds) {
+		tflog.Error(ctx, "failed to mutate Team roles", map[string]interface{}{"error": err})
+		diags.AddError(
+			"Unable to mutate Team roles, not every deployment role has a corresponding workspace role",
+			"Please ensure that every deployment role has a corresponding workspace role",
+		)
+		return diags
+	}
+	return diags
+}
+
+// CheckDuplicateWorkspaceId checks if there are duplicate workspace ids in the workspace roles
+func CheckDuplicateWorkspaceId(workspaceRoles []iam.WorkspaceRole) []string {
+	workspaceIdCount := make(map[string]int)
+	for _, role := range workspaceRoles {
+		workspaceIdCount[role.WorkspaceId]++
+	}
+
+	var duplicates []string
+	for id, count := range workspaceIdCount {
+		if count > 1 {
+			duplicates = append(duplicates, id)
+		}
+	}
```

Comment:
we can just do lo.UniqBy and then make sure the lengths are equal

Reply:
i did it this way so we can return what the duplicate ids are in the error message